resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala

```diff
@@ -1331,7 +1331,11 @@ private[spark] class Client(
       logInfo(s"Application report for $appId (state: $state)")
       logInfo(formatReportDetails(report, getDriverLogsLink(report)))
       if (state == YarnApplicationState.FAILED || state == YarnApplicationState.KILLED) {
+        cleanupStagingDir()
         throw new SparkException(s"Application $appId finished with status: $state")
+      } else {
+        val fs = stagingDirPath.getFileSystem(hadoopConf)
+        fs.deleteOnExit(stagingDirPath)
```

Comment:
Where is this otherwise cleaned up - is it also using deleteOnExit?

Reply:
> Where is this otherwise cleaned up - is it also using deleteOnExit?

1. If submitApplication failed, will clean staging https://github.com/apache/spark/blob/cc7cb7a803d5de03c526480c8968bbb2c3e82484/resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala#L231

2. When monitoring application, if ApplicationNotFound, will clean https://github.com/apache/spark/blob/cc7cb7a803d5de03c526480c8968bbb2c3e82484/resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala#L1168-L1171
3. When monitoring application, if Application end, will clean https://github.com/apache/spark/blob/cc7cb7a803d5de03c526480c8968bbb2c3e82484/resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala#L1214-L1218

Reply:
If we set `spark.yarn.submit.waitAppCompletion=false`, here `fireAndForget=true`
https://github.com/apache/spark/blob/cc7cb7a803d5de03c526480c8968bbb2c3e82484/resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala#L1326-L1336

If `launcherBackend ` just lose connection after `submitApplication`, means application success or failed quickly.
Here we won't clean staging dir in current code, In this pr, there are two case:
1. application just failed or killed, I think we need clean the staging dir directly.
2. For other status, since client won't wait application complete. The client just exit, we need to find other method to resolve this.

Reply:
> Where is this otherwise cleaned up - is it also using deleteOnExit?

Remove this code, it's not correct.

Reply:
Yeah either way there's a possible issue - what if cleaning the staging dir fails now?
It might be overkill, but, we could make sure that if cleanup fails, it still proceeds. try-catch here or in the clean method?

Reply:
> Yeah either way there's a possible issue - what if cleaning the staging dir fails now? It might be overkill, but, we could make sure that if cleanup fails, it still proceeds. try-catch here or in the clean method?

In clean method, there is already an ry-catch. Since I have seen some  `unregister` failed, so here change the order is necessary.